### PR TITLE
2620 3595 Writer backend selector, deprecating nifti_saver/writer, png_saver/writer

### LINKED ISSUE #2620: Image writer with itk backend to support more output image formats
**Is your feature request related to a problem? Please describe.**
The `SaveImage` transform should be extended to support writing output in different formats using ITK.
https://github.com/Project-MONAI/MONAI/blob/830f8bb34532f63e188b43738f804fff6f059d5b/monai/transforms/io/array.py#L260-L261

the current nifti output doesn't work well with Slicer https://github.com/Project-MONAI/MONAILabel/issues/211

see also https://github.com/InsightSoftwareConsortium/ITK/issues/2663

### LINKED ISSUE #3595: refactoring the image writer modules
This is a follow-up on the image writer of [the previous I/O module proposals](https://github.com/Project-MONAI/MONAI/issues/856).

### Goal and anticipated benefits
The current image writing module has been primarily built around the NIfTI specifications using the nibabel package as the backend.
The goal is to redesign the module in order to decouple the universal and the backend-specific implementations.

This will make it easier to bring in various writer backends such as ITK-Python, therefore supporting different image formats and more powerful writer module customisations.

related tickets: https://github.com/Project-MONAI/MONAI/issues/2620 https://github.com/Project-MONAI/MONAI/issues/2613 https://github.com/Project-MONAI/MONAILabel/issues/211
proof of concept https://github.com/Project-MONAI/MONAI/pull/3443

### Details
The writer module should have (1) universal logic (2) backend-specific logic (3) mechanism for defaulting/selecting backends. Specifically:

(1) In the context of image writing, the universal logic handles the image-related outputs from the deep learning workflows, such as:
-  fetching data array and metadata
-  arrangements/resampling of spatial/channel dimensions
-  preparing target data types
- preparing output folder structures, potentially compatible with BIDS (brain imaging data structure) and NDWB (neurodata without boarders) 

(2) The backend-specific logic includes
- creating data representations according to the backend APIs
- calling the backend APIs to finish the data writing

(3) The backend selection logic operates based on the user-specified parameters, backend system availability, and current system default configurations.

### PR DESCRIPTION
**Pull request by wyli**
closes #3595
closes #2620 

### Description
this is to add a `resolve_writer` and `register_writer`, 
update `monai.transform.SaveImage` to use the latest API. 

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

**Review by ericspod — APPROVED** [review]
Looks good to me!

File: monai/data/image_writer.py [hunk 1]
```diff
@@ -41,7 +42,69 @@
...
+            As an indexing key, it will be converted to a lower case string.
+        im_writer: one or multiple ImageWriter classes with high priority ones first.
+    """
+    fmt = f"{ext_name}".lower()
+    existing = look_up_option(fmt, SUPPORTED_WRITERS, default=())
+    all_writers = im_writer + existing
+    SUPPORTED_WRITERS[fmt] = all_writers
+
+
+def resolve_writer(ext_name, error_if_not_found=True) -> Sequence:
```
Comment (Nic-Ma) [inline monai/data/image_writer.py:82]: I feel the `resolve_writer` is confusing at first look, actually it doesn't "resolve" thing, maybe `get_writer` is more clear?

Thanks.

File: monai/data/image_writer.py [hunk 2]
```diff
@@ -716,3 +782,15 @@ def create_backend_obj(
             data = np.moveaxis(data, 0, 1)
 
         return PILImage.fromarray(data, mode=kwargs.pop("image_mode", None))
+
+
+def init():
+    """
+    Initialize the image writer modules according to the filename extension.
+    """
+    for ext in (".png", ".jpg", ".jpeg", ".bmp", ".tiff", ".tif"):
```
Comment (Nic-Ma) [inline monai/data/image_writer.py:791]: Here the `ext` contains "." at the beginning, while in readers we don't contain it:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/data/image_reader.py#L393
I feel it's better to be consistent?

Thanks.

File: monai/data/image_writer.py [hunk 3]
```diff
@@ -41,7 +42,69 @@
...
+    The customised writers could be registered by:
+
+    .. code-block:: python
+
+        from monai.data import image_writer
+        # `MyWriter` must implement `ImageWriter` interface
+        image_writer.register_writer(".nii", MyWriter)
+
+    Args:
+        ext_name: the filename extension of the image.
```
Comment (Nic-Ma) [inline monai/data/image_writer.py:72]: Here is arg is `ext_name` while in readers it's `suffixes`, I feel it's better to be consistent?

Thanks.

File: monai/transforms/io/array.py [hunk 4]
```diff
@@ -227,125 +227,152 @@ def __call__(self, filename: Union[Sequence[PathLike], PathLike], reader: Option
...
+            makedirs=True,
+            data_root_dir=data_root_dir,
+        )
+
+        self.output_ext = output_ext.lower() or output_format.lower()
+        self.writers = image_writer.resolve_writer(self.output_ext) if writer is None else (writer,)
+        self.writer_obj = None
+
+        _output_dtype = output_dtype
+        if self.output_ext == ".png" and _output_dtype not in (np.uint8, np.uint16):
```
Comment (Nic-Ma) [inline monai/transforms/io/array.py:317]: Maybe we should unify the `ext` and `.XXX` with readers `suffixes` and `XXX`?

Thanks.

File: monai/transforms/io/array.py [hunk 5]
```diff
@@ -227,125 +227,152 @@ def __call__(self, filename: Union[Sequence[PathLike], PathLike], reader: Option
...
+            _output_dtype = np.uint8
+        self.init_kwargs = {"output_dtype": _output_dtype, "scale": scale}
+        self.data_kwargs = {"squeeze_end_dims": squeeze_end_dims, "channel_dim": 0}
+        self.meta_kwargs = {"resample": resample, "mode": mode, "padding_mode": padding_mode, "dtype": dtype}
+        self.write_kwargs = {"verbose": print_log}
+        self._data_index = 0
+
+    def set_options(self, init_kwargs=None, data_kwargs=None, meta_kwargs=None, write_kwargs=None):
+        """
+        Set the options for the underlying writer by updating `kwargs` dictionaries.
```
Comment (Nic-Ma) [inline monai/transforms/io/array.py:329]: Maybe it can be more clear: updating the default `kwargs` which were generated in the constructor?

Thanks.

**Comment by Nic-Ma** [comment]
Hi @wyli ,

Thanks for the quick implementation, put some comments inline, I will review the `unit tests` later.

Thanks.

File: monai/data/image_writer.py [hunk 6]
```diff
@@ -41,7 +42,69 @@
...
+            As an indexing key, it will be converted to a lower case string.
+        im_writer: one or multiple ImageWriter classes with high priority ones first.
+    """
+    fmt = f"{ext_name}".lower()
+    existing = look_up_option(fmt, SUPPORTED_WRITERS, default=())
+    all_writers = im_writer + existing
+    SUPPORTED_WRITERS[fmt] = all_writers
```
[the remark this answers is not in the document]
Comment (wyli) [reply monai/data/image_writer.py:81]: `SUPPORTED_WRITERS ` is a module-level variable, it's not a global one. I believe it's a 
 very common design, for example, scikit-image and PIL used it in a similar way 
- https://github.com/scikit-image/scikit-image/blob/ed642e2bc822f362504d24379dee94978d6fa9de/skimage/io/manage_plugins.py#L39
- https://github.com/python-pillow/Pillow/blob/f5fab326fac3481308fc22349b869b691826e313/src/PIL/Image.py#L195

**Pull request merged**